architecture reference docs (#285)
cylc/cylc-doc: src/reference/architecture/ui-server.rst

See :ref:`managing-multiple-extensions` for details on managing which
extensions are run by `Jupyter Server`_.

`Jupyter Server`_ can be run in two ways, single-user (token authenticated)
and multi-user (hub authenticated).

.. _single-user mode:

Single-User Mode
----------------

In single-user mode users must start their own UI Servers from the command line.

`Jupyter Server`_ will provide them with a URL to access their server including
a secure token which provides authentication.

.. admonition:: Authentication Overview
   :class: hint

   See :ref:`server_security`:

Users can only monitor and control their own workflows.

.. image:: img/gui-arch-single-user.svg
   :width: 70%
   :align: center

.. _multi-user mode:

Multi-User Mode
----------------

Multi-user mode requires `Jupyter Hub`_ to be installed.

An administrator must start `Jupyter Hub`_ under a user account with
the required privileges to spawn UI Servers on behalf of the user.

Users then visit `Jupyter Hub`_ where they authenticate. `Jupyter Hub`_
spawns UI Servers on behalf of users and provides each with a fixed URL
(derived from the user name) using a `Configurable HTTP Proxy`_.

> Users can access each others UI Servers providing that they have been granted
> permissions.

Reviewer: ```suggestion
Users can access each other's UI Servers providing they have been granted
permission.
```

Authorisation is performed by Cylc within the `Cylc UI Server`_, see
:ref:`cylc authorisation`.

.. _Jupyter Hub technical overview: https://jupyterhub.readthedocs.io/en/stable/reference/technical-overview.html

For information on the architecture of `Jupyter Hub`_ and the
`Configurable HTTP Proxy`_ see the `Jupyter Hub technical overview`_.

.. admonition:: Authentication Overview
   :class: hint

   .. _Security In Jupyter Hub: https://jupyterhub.readthedocs.io/en/stable/reference/websecurity.html

   See `Security In Jupyter Hub`_.

.. image:: img/gui-arch-multi-user.svg
   :width: 100%

.. _cylc authorisation:

Authorisation
-------------

> *Authentication* is provided by either `Jupyter Server`_ or `Jupyter Hub`_.

Author: Could do, they do both build on the Tornado base classes.

> *Authorisation* in the `Cylc UI Server`_ is provided by Cylc, this is used in
> multi-user mode to allow users to connect to each others UI Servers for
> monitoring and or control purposes.

Reviewer: ```suggestion
*Authorisation* in the `Cylc UI Server`_ is provided by Cylc. In
multi-user mode this allows users to connect to each others UI Servers for
monitoring or control purposes.
```

.. note::

   This authorisation only applies to the ``/cylc`` endpoints managed by the
   Cylc extension.  Other `Jupyter Server`_ extensions such as `Jupyter Lab`_
   do not share this authorisation and will not be accessible to other users
   (unless they can be configured to allow this).

.. warning::

   Authorisation has not yet been implemented, in multi-user mode users can

>    only access their own UI Servers.

Reviewer 2: 👍 @datamel, we will have to remember to update it when your PR's are merged I think :-)

>    Authorisation has not yet been implemented, in multi-user mode users can
>    only access their own UI Servers.

Reviewer: ```suggestion
   Authorisation has not yet been fully implemented. For the moment, users can
   only access their own UI Servers in multi-user mode.
```
